desc/compute/_stability.py

```diff
@@ -309,115 +312,111 @@ def _ideal_ballooning_lambda(params, transforms, profiles, data, **kwargs):
-
+    f = (constant1 / (B**3 * gradpar)) * gds2
+    g = (constant2 / B * gradpar) * gds2
+    g_half = (g[..., 1:] + g[..., :-1]) / 2
     c = (
-        a_N**3
-        * B_N
-        * jnp.reshape(
-            2
-            / B_sup_zeta[None, ...]
-            * sign_psi
-            * rho**2
-            * dpdpsi
-            * (cvdrift - shear / (2 * rho**2) * zeta0[:, None, None, None] * cvdrift0),
-            (N_alpha, N_zeta0, N_zeta),
-        )
+        _dim(constant2 * sign_psi * dp_dpsi)
+        / B_sup_zeta
+        * (_dim(2 * rho**2) * cvdrift - (_dim(shear) * cvdrift0) * zeta0)
     )
-
-    h = phi[1] - phi[0]
-
-    i = jnp.arange(N_alpha)[:, None, None, None]
-    l = jnp.arange(N_zeta0)[None, :, None, None]
-    j = jnp.arange(N_zeta - 2)[None, None, :, None]
-    k = jnp.arange(N_zeta - 2)[None, None, None, :]
-
-    A = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
-    B_inv = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
-
-    A = A.at[i, l, j, k].set(
-        g_half[i, l, k] / h**2 * (j - k == -1)
-        + (-(g_half[i, l, j + 1] + g_half[i, l, j]) / h**2 + c[i, l, j + 1])
-        * (j - k == 0)
-        + g_half[i, l, j] / h**2 * (j - k == 1)
+    assert (
+        gds2.shape
+        == f.shape
+        == g.shape
+        == c.shape
+        == (zeta0.size, grid.num_rho, grid.num_alpha, grid.num_zeta)
     )
 
-    B_inv = B_inv.at[i, l, j, k].set(1 / jnp.sqrt(f[i, l, j + 1]) * (j - k == 0))
-
-    A_redo = B_inv @ A @ jnp.transpose(B_inv, axes=(0, 1, 3, 2))
+    j = jnp.arange(grid.num_zeta - 2)
+    diag_inner = c[..., 1:-1] - (g_half[..., 1:-1] + g_half[..., 2:]) / dz**2
+    diag_outer = g_half[..., 1:-2] / dz**2
+    A = jnp.zeros(
+        (zeta0.size, grid.num_rho, grid.num_alpha, grid.num_zeta - 2, grid.num_zeta - 2)
+    )
+    A = A.at[..., j, j].set(diag_inner)
+    A = A.at[..., j[:-1], j[1:]].set(diag_outer)
+    A = A.at[..., j[1:], j[:-1]].set(diag_outer)
 
     # TODO: Issue #1750
     # Try jax.scipy.eigh_tridiagonal or a better solver for improved performance
-    w, v = jnp.linalg.eigh(A_redo)
-
-    # Find the top_k eigenvalues.
-    top_eigvals, top_theta_idxs = jax.lax.top_k(w, k=Neigvals)
-
-    # v becomes less than the machine precision at some theta points which gives NaNs
-    # stop_gradient prevents that. Not sure how it will affect an objective that
-    # requires both the eigenvalue and eigenfunction
-    v_T = jnp.transpose(jax.lax.stop_gradient(v), axes=(0, 1, 3, 2))
-
-    top_eigfuns = jnp.take_along_axis(v_T, top_theta_idxs[..., None], axis=2)
-
-    data["ideal ballooning lambda"] = top_eigvals
-    data["ideal ballooning eigenfunction"] = top_eigfuns
+    if eigfuns:
+        B_inv = jnp.reciprocal(jnp.sqrt(f[..., 1:-1]))
+        w, v = jnp.linalg.eigh(B_inv[..., jnp.newaxis] * A * B_inv[..., jnp.newaxis, :])
+    else:
+        # We may seek eigenvalues of B⁻¹ @ A because for diagonal matrices,
```

Comment:
don't you still need the other factor of Binv to make the matrix symmetric (since eigvalsh only accesses one half of the matrix)